// matreshka/array/_pseudonativemethods/createsortingmethod.js
'use strict';

var initMK = require('../../_core/init');

var reportModified = require('../_reportmodified');

// creates sorting method and returns it (sort, reverse, sort_, reverse_)
module.exports = createSortingMethod;
function createSortingMethod(name, hasOptions) {
    return function pseudoNativeMethod(sortCallback) {
        if (this.length < 2) return this;
        initMK(this);

        var givenEventOptions = hasOptions ? arguments[arguments.length - 1] : null;
        var method = Array.prototype[name];

        var eventOptions = {
            method: name,
            self: this,
            added: [],
            removed: []
        };

        // call original method
        if (name === 'sort' && typeof sortCallback === 'function') {
            method.call(this, sortCallback);
        } else {
            method.call(this);
        }

        // extend event options by custom event options if they are given
        if (hasOptions) {
            if (givenEventOptions && typeof givenEventOptions === 'object') {
                var _result = eventOptions;

                for (var _source2 = givenEventOptions, _keys2 = Object.keys(_source2), _l2 = _keys2.length, _i2 = 0, _key2; _i2 < _l2; _i2++) {
                    _key2 = _keys2[_i2];
                    _result[_key2] = _source2[_key2];
                }
            }
        }

        reportModified(this, eventOptions);

        return this;
    };
}
