// seemple/array/_processrendering/processunshift.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = processUnshift;

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _renderitemnode = _interopRequireDefault(require("./renderitemnode"));

var _triggerone = _interopRequireDefault(require("../../trigger/_triggerone"));

var _checkalreadyrendered = _interopRequireDefault(require("./checkalreadyrendered"));

// this function renders inserted items if possible when unshift or push method is called
function processUnshift(_ref) {
  var self = _ref.self,
      selfDef = _ref.selfDef,
      eventOptions = _ref.eventOptions,
      container = _ref.container;
  var added = eventOptions.added,
      silent = eventOptions.silent; // iterate over all added items in opposite order

  for (var i = added.length - 1; i + 1; i--) {
    var item = added[i];

    if (item && (0, _typeof2["default"])(item) === 'object') {
      // if a node of an item is already rendered then throw an error
      (0, _checkalreadyrendered["default"])({
        item: item,
        selfDef: selfDef
      });

      var _renderItemNode = (0, _renderitemnode["default"])({
        selfDef: selfDef,
        self: self,
        item: item,
        eventOptions: eventOptions
      }),
          node = _renderItemNode.node,
          itemEventOptions = _renderItemNode.itemEventOptions;

      if (node) {
        if (container.firstChild) {
          container.insertBefore(node, container.firstChild);
        } else {
          container.appendChild(node);
        }

        if (!silent) {
          (0, _triggerone["default"])(item, 'afterrender', itemEventOptions);
        }
      }
    }
  }
}
