// matreshka/object/_prototype.js
'use strict';

var _afterInit$setData$ad;

var _afterInit = require('./_afterinit');

var addDataKeys = require('./adddatakeys');

var removeDataKeys = require('./removedatakeys');

var isDataKey = require('./isdatakey');

var setData = require('./setdata');

var keyOf = require('./keyof');

var keys = require('./keys');

var values = require('./values');

var entries = require('./entries');

var toJSON = require('./tojson');

var each = require('./each');

var iterator = require('./iterator');

var symbolIterator = typeof Symbol === 'function' ? Symbol.iterator : '@@iterator';

module.exports = (_afterInit$setData$ad = {
    _afterInit: _afterInit,
    setData: setData,
    addDataKeys: addDataKeys,
    removeDataKeys: removeDataKeys,
    isDataKey: isDataKey,
    keys: keys,
    values: values,
    entries: entries,
    keyOf: keyOf,
    toJSON: toJSON,
    each: each,
    isMatreshkaObject: true,
    jset: setData }, _afterInit$setData$ad[symbolIterator] = iterator, _afterInit$setData$ad);
