// file-binders/file.js
"use strict";

var getFileReaderMethodName = require('./_get-filereader-method-name');

var readFiles = require('./_read-files');

function createFileChangeHandler(_ref) {
  var callback = _ref.callback,
      methodName = _ref.methodName;
  return function fileChangeHandler() {
    var files = this.files;

    if (files.length) {
      readFiles(files, methodName, callback);
    } else {
      callback([]);
    }
  };
}

module.exports = function fileBinder(readAs) {
  var methodName = readAs ? getFileReaderMethodName(readAs) : null;
  var fileChangeHandler;
  return {
    on: function on(callback) {
      fileChangeHandler = createFileChangeHandler({
        callback: callback,
        methodName: methodName
      });
      this.addEventListener('change', fileChangeHandler);
    },
    destroy: function destroy() {
      this.removeEventListener('change', fileChangeHandler);
    },
    getValue: function getValue(_ref2) {
      var domEvent = _ref2.domEvent;
      var files = domEvent || [];
      return this.multiple ? files : files[0] || null;
    },
    setValue: null
  };
};
