// matreshka/_dom/mq/_html2nodelist.js
'use strict';

// converts HTML string to NodeList instance
module.exports = html2nodeList;
function html2nodeList(givenHTML) {
    // wrapMap is taken from jQuery
    var wrapMap = {
        option: [1, '<select multiple="multiple">', '</select>'],
        legend: [1, '<fieldset>', '</fieldset>'],
        thead: [1, '<table>', '</table>'],
        tr: [2, '<table><tbody>', '</tbody></table>'],
        td: [3, '<table><tbody><tr>', '</tr></tbody></table>'],
        col: [2, '<table><tbody></tbody><colgroup>', '</colgroup></table>'],
        area: [1, '<map>', '</map>'],
        _: [0, '', '']
    };

    var html = givenHTML.replace(/^\s+|\s+$/g, '');
    var node = window.document.createElement('div');
    var i = void 0;

    wrapMap.optgroup = wrapMap.option;
    wrapMap.tbody = wrapMap.tfoot = wrapMap.colgroup = wrapMap.caption = wrapMap.thead;
    wrapMap.th = wrapMap.td;

    var ex = /<([\w:]+)/.exec(html);
    var wrapper = ex && wrapMap[ex[1]] || wrapMap._;

    node.innerHTML = wrapper[1] + html + wrapper[2];

    i = wrapper[0];

    while (i) {
        i -= 1;
        node = node.children[0];
    }

    return node.childNodes;
}
