// seemple/array/_pseudonativemethods/createsortingmethod.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = createSortingMethod;

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _init = _interopRequireDefault(require("../../_core/init"));

var _reportmodified = _interopRequireDefault(require("../_reportmodified"));

var _assign = _interopRequireDefault(require("../../_helpers/assign"));

// creates sorting method and returns it (sort, reverse, sort_, reverse_)
function createSortingMethod(name, hasOptions) {
  return function pseudoNativeMethod(sortCallback) {
    if (this.length < 2) return this;
    (0, _init["default"])(this);
    var givenEventOptions = hasOptions ? arguments[arguments.length - 1] : null;
    var method = Array.prototype[name];
    var eventOptions = {
      method: name,
      self: this,
      added: [],
      removed: []
    }; // call original method

    if (name === 'sort' && typeof sortCallback === 'function') {
      method.call(this, sortCallback);
    } else {
      method.call(this);
    } // extend event options by custom event options if they are given


    if (hasOptions) {
      if (givenEventOptions && (0, _typeof2["default"])(givenEventOptions) === 'object') {
        (0, _assign["default"])(eventOptions, givenEventOptions);
      }
    }

    (0, _reportmodified["default"])(this, eventOptions);
    return this;
  };
}
