// matreshka/parsebindings/_parserdata.js
'use strict';

var calc = require('../calc');

var parserBrackets = require('../parserbrackets');

var parserData = {};

// since Matreshka allows to change parser brackets via parserBrackets objects
// the parser needs to generate required regular expressions and escaped brackets every time
// when parseBindings is called
// to optimize this behavior parserData object is created
// it calculates needed data every time when parserBrackets are changed
// and when parseBindings function is called it uses previously generated regeps
// from parserData object
calc(parserData, {
    leftBracket: {
        source: {
            object: parserBrackets,
            key: 'left'
        }
    },
    rightBracket: {
        source: {
            object: parserBrackets,
            key: 'right'
        }
    },
    escLeftBracket: {
        source: 'leftBracket',
        handler: function (left) {
            return left.replace(/(\[|\(|\?)/g, '\\$1');
        }
    },
    escRightBracket: {
        source: 'rightBracket',
        handler: function (right) {
            return right.replace(/(]|\)|\?)/g, '\\$1');
        }
    },
    bindingReg: {
        source: ['escLeftBracket', 'escRightBracket'],
        handler: function (left, right) {
            return new RegExp(left + '\\s*(.+?)\\s*' + right, 'g');
        }
    },
    strictBindingReg: {
        source: ['escLeftBracket', 'escRightBracket'],
        handler: function (left, right) {
            return new RegExp('^' + left + '\\s*(.+?)\\s*' + right + '$', 'g');
        }
    }
}, {
    debounceCalc: false // we need to get new regexps immediately when brackets are changed
});

module.exports = parserData;
