// matreshka/array/_processrendering/processspliceadd.js
'use strict';

var renderItemNode = require('./renderitemnode');

var triggerOne = require('../../trigger/_triggerone');

var checkAlreadyRendered = require('./checkalreadyrendered');

var getAlreadyRendered = require('./getalreadyrendered');

// the function handles rendering of added items passed as third and rest arguments to splice method
module.exports = processSpliceAdd;
function processSpliceAdd(_ref) {
    var self = _ref.self,
        selfDef = _ref.selfDef,
        eventOptions = _ref.eventOptions,
        container = _ref.container;
    var added = eventOptions.added,
        silent = eventOptions.silent;

    var nextIndex = self.lastIndexOf(added[added.length - 1]) + 1;
    var next = self[nextIndex];
    var nextNode = void 0;

    // get a node of an item which is placed next to the last added item
    // it is needed to insert newly rendered items before
    if (next && typeof next === 'object') {
        nextNode = getAlreadyRendered({
            item: next,
            selfDef: selfDef
        });
    }

    for (var _target = added, _index = 0, item, _l = _target.length; item = _target[_index], _index < _l; _index++) {
        if (item && typeof item === 'object') {
            // throw an error if node of an item is alread rendered
            checkAlreadyRendered({
                item: item,
                selfDef: selfDef
            });

            var _renderItemNode = renderItemNode({
                selfDef: selfDef,
                self: self,
                item: item,
                eventOptions: eventOptions
            }),
                node = _renderItemNode.node,
                itemEventOptions = _renderItemNode.itemEventOptions;

            if (node) {
                if (nextNode) {
                    container.insertBefore(node, nextNode);
                } else {
                    container.appendChild(node);
                }

                if (!silent) {
                    triggerOne(item, 'afterrender', itemEventOptions);
                }
            }
        }
    }
}
